### DISCUSSION TITLE
Collection Initializer could initialize collections with a capacity

### DISCUSSION THREAD

**Discussion by xtqqczze** [dotnet/csharplang#4930] [Language Ideas]
See: https://github.com/dotnet/roslyn/issues/2628

**Comment by ZacharyPatten** [comment]
~~It would be nice if there was an attribute to map the "capacity" to a parameter of a different name. What if I write a custom data structure that has `minimumCapacity` or `initialCapacity`, or anything other than just `capacity`? That seems like a case where an attribute could help so I don't have to call the parameter `capacity` but can still have optimized initializations.~~

**Reply by bernd5** [reply]
Why not just calling the appropiate constructor?

**Reply by ZacharyPatten** [reply]
My comment was only if this issue was added to the language. (I should have clarified).

The idea behind the issue is to optimize code during compile time if the programmer did _not_ use the correct constructor.

But... thinking about it more than two seconds... this is issue is probably not a good idea. **I think I would just prefer analyzers and compiler warning rather than changing the compilation itself.** Just show a warning if the code could be optimized by using a different constructor. Someone has probably already written an analyzer for this topic.

**Comment by canton7** [comment]
Also related: dotnet/roslyn#43944. See Cyrus' comments there.

**Comment by HaloFour** [comment]
I'd rather consider whether something like this would fit with target-typed list literals that are being considered as a counterpoint to list patterns/deconstruction.

```cs
List<int> list = [ 1, 2, 3, 4, 5 ];
```
